src/main/java/io/appium/java_client/plugins/storage/StorageClient.java

```diff
@@ -0,0 +1,248 @@
+import org.openqa.selenium.remote.http.HttpHeader;
+import org.openqa.selenium.remote.http.HttpMethod;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+import org.openqa.selenium.remote.http.WebSocket;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static io.appium.java_client.plugins.storage.StorageUtils.calcSha1Digest;
+import static io.appium.java_client.plugins.storage.StorageUtils.streamFileToWebSocket;
+
+/**
+ * This is a Java implementation of the Appium server storage plugin client.
+ * See <a href="https://github.com/appium/appium/blob/master/packages/storage-plugin/README.md">the plugin README</a>
+ * for more details.
+ */
+public class StorageClient {
+    public static final String PREFIX = "/storage";
+    private final Json json = new Json();
+    private final AbstractHttpResponseCodec responseCodec = new W3CHttpResponseCodec();
+    private final ErrorCodec errorCodec = ErrorCodec.createDefault();
+
+    private final URL baseUrl;
+    private final HttpClient httpClient;
+
+    public StorageClient(URL baseUrl) {
+        this.baseUrl = baseUrl;
+        this.httpClient = HttpClient.Factory.createDefault().createClient(baseUrl);
+    }
+
+    public StorageClient(ClientConfig clientConfig) {
+        this.httpClient = HttpClient.Factory.createDefault().createClient(clientConfig);
+        this.baseUrl = clientConfig.baseUrl();
+    }
+
+    /**
+     * Adds a local file to the server storage.
+     * The remote file name is be set to the same value as the local file name.
+     *
+     * @param file File instance.
+     */
+    public void add(File file) {
+        add(file, file.getName());
+    }
+
+    /**
+     * Adds a local file to the server storage.
+     *
+     * @param file File instance.
+     * @param name The remote file name.
+     */
+    public void add(File file, String name) {
+        var request = new HttpRequest(HttpMethod.POST, formatPath(baseUrl, PREFIX, "add").toString());
+        var httpResponse = httpClient.execute(setJsonPayload(request, Map.of(
+                "name", name,
+                "sha1", calcSha1Digest(file)
+        )));
+        Map<String, Object> value = requireResponseValue(httpResponse);
+        final var wsTtlMs = (Long) value.get("ttlMs");
+        //noinspection unchecked
+        var wsInfo = (Map<String, Object>) value.get("ws");
+        var streamWsPathname = (String) wsInfo.get("stream");
+        var eventWsPathname = (String) wsInfo.get("events");
+        final var completion = new CountDownLatch(1);
```

Comment:
Could `CompletableFuture` be good for modern Java instead of `CountDownLatch`? I don't have much experience this in Java, but I saw CompletableFuture usage for async before

Reply:
I would say everything is good that does its job well, and is not deprecated, and is readable :)

The concept of CompletableFuture does not make java code async, it just hides the threading complexity under another abstraction layer.
If you have a proposal on how to make the same implementation shorter/better understable via CompletableFuture use then I would be happy to apply it